pricegraph/src/orderbook.rs

```diff
@@ -818,14 +823,65 @@ mod tests {
         let overlap = orderbook.reduce_overlapping_transitive_orderbook(1, 2);
 
         // Transitive order `1 -> 2` buying 1 selling 2
+        assert_eq!(overlap.asks.len(), 1);
         assert_approx_eq!(overlap.asks[0].buy, 1_000_000.0);
         assert_approx_eq!(overlap.asks[0].sell, 2_000_000.0);
 
         // Transitive order `2 -> 3 -> 1` buying 2 selling 1
+        assert_eq!(overlap.bids.len(), 1);
         assert_approx_eq!(overlap.bids[0].buy, 500_000.0);
         assert_approx_eq!(overlap.bids[0].sell, 500_000.0 / FEE_FACTOR);
     }
 
+    #[test]
+    fn includes_transitive_order_only_once() {
+        // /---0.5---v
+        // 0         1
+        // ^---1.0---/
+        // ^---1.5--/
+        let mut orderbook = orderbook! {
+            users {
+                @1 {
+                    token 1 => 100_000_000,
+                }
+                @2 {
+                    token 0 => 1_000_000,
+                }
+                @3 {
+                    token 0 => 1_000_000,
+                }
+            }
+            orders {
+                owner @1 buying 0 [50_000_000] selling 1 [100_000_000],
+                owner @2 buying 1 [1_000_000] selling 0 [1_000_000],
+                owner @3 buying 1 [1_500_000] selling 0 [1_000_000],
+            }
+        };
+
+        let overlap = orderbook.reduce_overlapping_transitive_orderbook(0, 1);
+
+        // Transitive order `0 -> 1` buying 0 selling 1
+        assert_eq!(overlap.asks.len(), 1);
+        assert_approx_eq!(overlap.asks[0].buy, 50_000_000.0);
+        assert_approx_eq!(overlap.asks[0].sell, 100_000_000.0);
+
+        // Transitive order `1 -> 0` buying 1 selling 0
+        assert_eq!(overlap.bids.len(), 1);
+        assert_approx_eq!(overlap.bids[0].buy, 1_000_000.0);
+        assert_approx_eq!(overlap.bids[0].sell, 1_000_000.0);
+
+        // NOTE: This is counter-intuitive, but there should only be one
+        // transitive order from `0 -> 1`. This is because whenever a negative
+        // cycle is found, it gets split into two transitive orders, one from
+        // `base -> quote` (ask) and the other from `quote -> base` (bid). These
+        // transitive orders are **completely** reduced, so even if there are
+        // more orders that could overlap with the remaining amount of a
+        // transitive order if only the negative cycle we reduced, they might
+        // not get found by `reduce_overlapping_transitive_orderbook`. Rest
```

Comment:
Something with this sentence is weird (it's also a bit too long for my brain stack)